File: scripts/dispatch_hook.py
#!/usr/bin/env python3
import sys
import os
import subprocess
import platform
import shutil
import json

def log_error(message):
    """Logs error to stderr and optionally to a file."""
    print(f"Dispatcher Error: {message}", file=sys.stderr)
    try:
        # Attempt to log to the extension's debug.log
        ext_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
        log_path = os.path.join(ext_dir, "debug.log")
        with open(log_path, "a") as f:
            from datetime import datetime
            f.write(f"[{datetime.now().isoformat()}] [dispatch_hook] {message}\n")
    except:
        pass

def main():
    if len(sys.argv) < 2:
        print("Usage: dispatch_hook.py <hook_name> [args...]", file=sys.stderr)
        sys.exit(1)

    hook_name = sys.argv[1]
    extra_args = sys.argv[2:]
    
    script_dir = os.path.dirname(os.path.abspath(__file__))
    ext_dir = os.path.dirname(script_dir)
    hooks_dir = os.path.join(ext_dir, "hooks")
    is_windows = "windows" in platform.system().lower()

    # Determine script path and command
    if is_windows:
        script_path = os.path.join(hooks_dir, f"{hook_name}.ps1")
        exe = shutil.which('pwsh') or shutil.which('powershell')
        if not exe:
            log_error("PowerShell not found.")
            print('{"decision": "allow"}')
            sys.exit(0)
        cmd = [exe, "-ExecutionPolicy", "Bypass", "-File", script_path] + extra_args
    else:
        script_path = os.path.join(hooks_dir, f"{hook_name}.sh")
        cmd = ["bash", script_path] + extra_args

    if not os.path.exists(script_path):
        log_error(f"Hook script not found: {script_path}")
        print('{"decision": "allow"}')
        sys.exit(0)

    # Read stdin safely
    input_data = ""
    if not sys.stdin.isatty():
        try:
            input_data = sys.stdin.read()
        except Exception as e:
            log_error(f"Failed to read stdin: {e}")

    try:
        # Execute the hook script
        # We pass env vars to help the script find the extension
        env = os.environ.copy()
        env["EXTENSION_DIR"] = ext_dir
        
        result = subprocess.run(
            cmd,
            input=input_data,
            text=True,
            capture_output=True,
            env=env
        )
        
        # Output handling
        if result.stdout:
            print(result.stdout, end='')
        
        if result.stderr:
            print(result.stderr, file=sys.stderr)
            
        # If stdout was empty, ensure we at least return allow
        if not result.stdout.strip():
            # If the script failed but didn't output anything, default to allow
            if result.returncode != 0:
                log_error(f"Hook {hook_name} failed with code {result.returncode} and no output.")
            print('{"decision": "allow"}')
            
        sys.exit(result.returncode)
        
    except Exception as e:
        log_error(f"Unexpected execution error: {e}")
        print('{"decision": "allow"}')
        sys.exit(0)

if __name__ == "__main__":
    main()

File: extension/dispatch_hook.js
#!/usr/bin/env node
import { spawn } from 'node:child_process';
import { existsSync, appendFileSync } from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
const __dirname = dirname(fileURLToPath(import.meta.url));
const EXTENSION_DIR = resolve(__dirname, '..');
const HOOKS_DIR = join(EXTENSION_DIR, 'hooks');
const LOG_PATH = join(EXTENSION_DIR, 'debug.log');
function logError(message) {
    console.error(`Dispatcher Error: ${message}`);
    try {
        const timestamp = new Date().toISOString();
        appendFileSync(LOG_PATH, `[${timestamp}] [dispatch_hook] ${message}
`);
    }
    catch { /* ignore */ }
}
function allow() {
    console.log(JSON.stringify({ decision: 'allow' }));
}
function findExecutable(name) {
    const pathEnv = process.env.PATH || '';
    const paths = pathEnv.split(process.platform === 'win32' ? ';' : ':');
    const extensions = process.platform === 'win32' ? ['.exe', '.cmd', '.bat', '.ps1', ''] : [''];
    for (const p of paths) {
        for (const ext of extensions) {
            const fullPath = join(p, name + ext);
            if (existsSync(fullPath))
                return fullPath;
        }
    }
    return null;
}
async function main() {
    const args = process.argv.slice(2);
    if (args.length < 1) {
        console.error('Usage: dispatch_hook <hook_name> [args...]');
        process.exit(1);
    }
    const [hookName, ...extraArgs] = args;
    const isWindows = process.platform === 'win32';
    let scriptPath;
    let cmd;
    let cmdArgs;
    if (isWindows) {
        scriptPath = join(HOOKS_DIR, `${hookName}.ps1`);
        const exe = findExecutable('pwsh') || findExecutable('powershell');
        if (!exe) {
            logError("PowerShell not found.");
            allow();
            process.exit(0);
        }
        cmd = exe;
        cmdArgs = ['-ExecutionPolicy', 'Bypass', '-File', scriptPath, ...extraArgs];
    }
    else {
        scriptPath = join(HOOKS_DIR, `${hookName}.sh`);
        cmd = 'bash';
        cmdArgs = [scriptPath, ...extraArgs];
    }
    if (!existsSync(scriptPath)) {
        logError(`Hook script not found: ${scriptPath}`);
        allow();
        process.exit(0);
    }
    let inputData = '';
    if (!process.stdin.isTTY) {
        try {
            const chunks = [];
            for await (const chunk of process.stdin) {
                chunks.push(chunk);
            }
            inputData = Buffer.concat(chunks).toString();
        }
        catch { /* ignore */ }
    }
    try {
        const child = spawn(cmd, cmdArgs, {
            env: { ...process.env, EXTENSION_DIR },
            stdio: ['pipe', 'pipe', 'pipe']
        });
        if (inputData)
            child.stdin?.write(inputData);
        child.stdin?.end();
        let stdout = '';
        let stderr = '';
        child.stdout?.on('data', data => stdout += data.toString());
        child.stderr?.on('data', data => stderr += data.toString());
        child.on('close', code => {
            if (stdout)
                process.stdout.write(stdout);
            if (stderr)
                process.stderr.write(stderr);
            if (!stdout.trim()) {
                if (code !== 0 && code !== null) {
                    logError(`Hook ${hookName} failed with code ${code} and no output.`);
                }
                allow();
            }
            process.exit(code ?? 0);
        });
        child.on('error', err => {
            logError(`Failed to start child process: ${err}`);
            allow();
            process.exit(0);
        });
    }
    catch (e) {
        logError(`Unexpected execution error: ${e}`);
        allow();
        process.exit(0);
    }
}
main();
